From 5793f81d8ff88ade0e22063a95a7efbf3006f61b Mon Sep 17 00:00:00 2001
From: "Thulasimani,Sivakumar" <sivakumar.thulasimani@intel.com>
Date: Mon, 21 Dec 2015 16:24:20 +0530
Subject: [PATCH 05/19] FOR_UPSTREAM [VPG]: drm/i915: Fix audio checks during
 modeset

CHT/VLV systems have LP audio where audio driver communicates
to the HW through i915, hence it is not necessary to enable
audio output in port during modeset for such configs. This also
avoids the underrun seen during modeset on such systems.

check_crtc_state check for audio capability programming if it
is supported, but since in CHT/VLV we can enable the audio
output bit when audio playback starts it is not set during
modeset. checking for a bit not explicity set (in case of VLV)
is not correct. hence this patch modifies this check to be
done for all platforms other than CHT/VLV

Change-Id: Id25e568dcaeb0a9ca0837a7b66c433da9f964ada
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-30099
Signed-off-by: Sivakumar Thulasimani <sivakumar.thulasimani@intel.com>
Reviewed-on: https://android.intel.com:443/498180
---
 drivers/gpu/drm/i915/intel_display.c |  3 ++-
 drivers/gpu/drm/i915/intel_dp.c      | 10 ++++++++--
 drivers/gpu/drm/i915/intel_hdmi.c    | 10 ++++++++--
 3 files changed, 18 insertions(+), 5 deletions(-)

diff --git a/drivers/gpu/drm/i915/intel_display.c b/drivers/gpu/drm/i915/intel_display.c
index e08bdb6..3f5ca99 100644
--- a/drivers/gpu/drm/i915/intel_display.c
+++ b/drivers/gpu/drm/i915/intel_display.c
@@ -12225,7 +12225,8 @@ intel_pipe_config_compare(struct drm_device *dev,
 	    IS_VALLEYVIEW(dev))
 		PIPE_CONF_CHECK_I(limited_color_range);
 
-	PIPE_CONF_CHECK_I(has_audio);
+	if (!IS_VALLEYVIEW(dev))
+		PIPE_CONF_CHECK_I(has_audio);
 
 	PIPE_CONF_CHECK_FLAGS(adjusted_mode.flags,
 			      DRM_MODE_FLAG_INTERLACE);
diff --git a/drivers/gpu/drm/i915/intel_dp.c b/drivers/gpu/drm/i915/intel_dp.c
index 1a516d1..bfadc22 100644
--- a/drivers/gpu/drm/i915/intel_dp.c
+++ b/drivers/gpu/drm/i915/intel_dp.c
@@ -1208,9 +1208,15 @@ static void intel_dp_prepare(struct intel_encoder *encoder)
 	intel_dp->DP |= DP_VOLTAGE_0_4 | DP_PRE_EMPHASIS_0;
 	intel_dp->DP |= DP_PORT_WIDTH(intel_dp->lane_count);
 
-	if (crtc->config.has_audio) {
+	/*
+	 * LP audio driver communicates through i915
+	 * where we enable audio output on port
+	 * hence not needed to be enabled during
+	 * modeset
+	 */
+	if (crtc->config.has_audio && !IS_VALLEYVIEW(dev)) {
 		DRM_DEBUG_DRIVER("Enabling DP audio on pipe %c\n",
-				 pipe_name(crtc->pipe));
+				pipe_name(crtc->pipe));
 		intel_dp->DP |= DP_AUDIO_OUTPUT_ENABLE;
 		intel_write_eld(&encoder->base, adjusted_mode);
 	}
diff --git a/drivers/gpu/drm/i915/intel_hdmi.c b/drivers/gpu/drm/i915/intel_hdmi.c
index d5ae6db..a5533ec 100644
--- a/drivers/gpu/drm/i915/intel_hdmi.c
+++ b/drivers/gpu/drm/i915/intel_hdmi.c
@@ -692,7 +692,13 @@ static void intel_hdmi_prepare(struct intel_encoder *encoder)
 	if (crtc->config.has_hdmi_sink)
 		hdmi_val |= HDMI_MODE_SELECT_HDMI;
 
-	if (crtc->config.has_audio) {
+	/*
+	 * LP audio driver communicates through i915
+	 * where we enable audio output on port
+	 * hence not needed to be enabled during
+	 * modeset
+	 */
+	if (crtc->config.has_audio && !IS_VALLEYVIEW(dev)) {
 		WARN_ON(!crtc->config.has_hdmi_sink);
 		DRM_DEBUG_DRIVER("Enabling HDMI audio on pipe %c\n",
 				 pipe_name(crtc->pipe));
@@ -787,7 +793,7 @@ static void intel_enable_hdmi(struct intel_encoder *encoder)
 	u32 temp;
 	u32 enable_bits = SDVO_ENABLE;
 
-	if (intel_crtc->config.has_audio)
+	if (intel_crtc->config.has_audio && !IS_VALLEYVIEW(dev))
 		enable_bits |= SDVO_AUDIO_ENABLE;
 
 	temp = I915_READ(intel_hdmi->hdmi_reg);
-- 
1.9.1

